### PR TITLE
Improve fake mode support by adding fake_context to ExportOutput

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105477
* __->__ #105247
* #105246

Prior to this PR, if the user called `fake_model.load_state_dict()` from within `enable_fake_mode`, the initial model state dict (including non persistent buffers) would not be reused by `ExportOutput.save` during ONNX proto creation.

That is not necessarily a bug because `ExportOutput.save` has a `model_state_dict` in which they can specify any state they want. However, it can be a hassle because if the user doesn't provide a full state, including non-persistent buffers, the resulting ONNX graph would require the missing buffers to be specified as input during execution. 

With this PR, the `enable_fake_mode` is improved to capture the initial model state including any non-persistent buffer. This reference (not actual data) is persisted within `ExportOutput` and used by `save` to load additional `state_dict` that was captured by `enable_fake_mode`. The result is an ONNX graph with all model state without user having to specify the non-persistent buffers.

This helps addressing https://github.com/pytorch/pytorch/issues/105233 for models that call `fake_model.load _state_dict` under the hood as potential buffers not returned by `model.state_dict()` may be captured.

ps: https://github.com/pytorch/pytorch/issues/105464 tracks pending tasks/limitations from this PR

### CODE REVIEW

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py [hunk 1]
```diff
@@ -931,15 +930,17 @@ def forward(self, x):
         def create_args():
             return (torch.rand(5, 2, 2),)
 
-        def create_pytorch_only_extra_kwargs():
```
Comment (BowenBao) [inline test/onnx/test_fx_to_onnx_with_onnxruntime.py:934]: How was this merged last time ...

File: torch/onnx/_internal/fx/serialization.py [hunk 2]
```diff
@@ -134,19 +134,22 @@ def save_model_with_external_data(
...
+            # This block tries to match the onnx initializer name with torch parameter/buffer
+            #  e.g. A pytorch buffer 'transformer.h.0.attn.bias' can be named 'h.0.attn.bias' in a ONNX initializer
+            # For each PyTorch tensor name loaded by torch.load,
+            #  1.  Search its best match in ONNX model. E.g., the match of
+            #       "transformer_attention_weight" could be "attention_weight".
+            #  2.  Set "tensor" as the initializer of the matched ONNX input.
+            #      E.g., "tensor" is stored as the initializer of "attention_weight".
+            # Step 1 is required because sometimes, tensor names are stored with prefix the dictionary
+            # loaded by torch.load.
+            for onnx_input_name in onnx_input_names:
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/serialization.py:147]: why is this change necessary?

**Review by BowenBao — COMMENTED** [review]
Hmm I'm wondering if the larger problem is about how to handle the tensor constants that are neither parameters nor persistent buffers under the model. A few questions

* Are non-persistent buffers the only case left for tensor constants that are not parameters or persistent buffers?
* What about only fakefying the things inside the checkpoint? That implies we set `allow_non_fake_inputs` and the issue is gone?

**Comment by BowenBao** [comment]
Can you help me point to the place where the non-persistent buffer data, that is not returned by `state_dict()`, is recorded?

File: test/onnx/test_fx_to_onnx_with_onnxruntime.py (hunk 1 above)
Comment (thiagocrepaldi) [reply test/onnx/test_fx_to_onnx_with_onnxruntime.py:934]: copy&paste not noticed during coding/review

**Comment by thiagocrepaldi** [comment]
> Hmm I'm wondering if the larger problem is about how to handle the tensor constants that are neither parameters nor persistent buffers under the model. A few questions
> 
> * Are non-persistent buffers the only case left for tensor constants that are not parameters or persistent buffers?
We will learn about corner cases and fix them as we use this feature more often. We didn't know about the non-persistent buffers until a few days ago, but now we have an official way to capture the non persistent buffer from checkpoint and a workaround for users to manually specify them (if ever needed)

> * What about only fakefying the things inside the checkpoint? That implies we set `allow_non_fake_inputs` and the issue is gone?
That issue is not tackled by this PR. Here all we care about is to capture a state when one is available. Before this PR, we discarded state that was given for free and then required the user to specify it. 

https://github.com/pytorch/pytorch/pull/105477 gives a first shot at that and already allow `allow_non_fake_inputs=False` for some scenarios

**Comment by BowenBao** [comment]
I see, so essentially `transformers.AutoModel.from_pretrained(model_name)` and `model.load_state_dict()` creates (or loads?) non-persistent buffer, and the `ONNXPytorchPatcher` hijacks and keeps track of it.

File: torch/onnx/_internal/fx/serialization.py (hunk 2 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/serialization.py:147]: IIRC, the `trasnformer.` prefix came from the checkpoint file, not the actual torch model names.

Comment (BowenBao) [reply torch/onnx/_internal/fx/serialization.py:147]: > Step 1 is required because sometimes, tensor names are stored with prefix the dictionary
> loaded by torch.load.

Hmm.. ok, so names in ckpt are *sometimes* different to names in `named_parameters` and `named_buffers`. I think there could be a more systematic fix then. Please also track this as one of the issues.

**Review by BowenBao — APPROVED** [review]
We should probably follow up on the naming prefix issue, others look good.

**Comment by thiagocrepaldi** [comment]
> We should probably follow up on the naming prefix issue, others look good.

Added https://github.com/pytorch/pytorch/issues/105751 to my task list to address this
